src/commands/exec.ts

```diff
@@ -85,22 +85,25 @@ export default class Exec extends BaseCommand {
     const ws = await this.getWebSocket(uri);
 
     await new Promise((resolve, reject) => {
-      const duplex = createWebSocketStream(ws, { encoding: 'utf-8' });
-      duplex.pipe(this.getOutputTransform());
-
-      if (flags.stdin) {
-        if (flags.tty) {
-          // This method is only available when stdin is a TTY as it's part of the tty.ReadStream class:
-          // https://nodejs.org/api/tty.html#readstreamsetrawmodemode
-          process.stdin.setRawMode(true);
-        }
-        process.stdin.pipe(this.getInputTransform()).pipe(duplex);
+      const websocket = createWebSocketStream(ws, { encoding: 'utf-8' });
+      const inputTransform = this.getInputTransform();
+      websocket.pipe(this.getOutputTransform());
+
+      if (process.stdin.isTTY) {
+        // This method is only available when stdin is a TTY as it's part of the tty.ReadStream class:
+        // https://nodejs.org/api/tty.html#readstreamsetrawmodemode
+        process.stdin.setRawMode(true);
+        process.stdin.pipe(inputTransform).pipe(websocket);
+      } else if (flags.stdin) {
+        // If stdin is not a tty, stdin can be closed before data gets received by the websocket.
+        // Passing {end: false} keeps the websocket open until it's closed when we receive the result.
+        process.stdin.pipe(inputTransform, { end: false }).pipe(websocket);
```

Comment:
Is it possible for us to write a test for this? Would be nice to cover both cases of tty and no-tty. If its going to be too difficult we can add a tech debt issue.

Reply:
It's maybe possible but very tricky because the tests in github actions run such that stdin isn't a tty, whereas running locally in your terminal via just `npm run test` does have a tty. So, it's probably not that hard to write a test that works locally for this, but not sure how valuable that is if it wouldn't work with the current way our tests are being run automatically on each branch